import matplotlib.pyplot as plt
import pandas as pd
import networkx as nx
import numpy as np
import urllib
from biomart import BiomartServer
import pickle
import glob
import random
import tqdm
import gseapy as gp
from gseapy import Biomart
import glob
import sys
import hubris_functions as hf

#we read in the GPickle generated by create_HUBRIS.py 
with open('G_hubris.gpickle', 'rb') as f:
    G_hubris= pickle.load(f)

#we only keep the largest connected component: 
G_hubris=G_hubris.subgraph(max(nx.connected_components(G_hubris), key=len))
print('N=',G_hubris.number_of_nodes(), 'E=',G_hubris.number_of_edges())

non_translated_nodes=[]
for i in G_hubris.nodes():
    if '_' in i:
        non_translated_nodes.append(i)
print('Nr of nodes with no translation:',len(non_translated_nodes))

#PARAMETERS:

cell_line='union' #options: intersection, union, IMR90, 16HBE
cell_lines=['IMR90','16HBE'] 
filter_crapome=True
filter_networks_based_on_expression=True

copd_gwas_genes=['FAM13A',
                 'IREB2',
                 'DSP',
                 'AGER',
                 'MFAP2',
                 'FBLN5',
                 'NPNT',
                 'FBXO38',
                 'SFTPD',
                 'TET2',
                 'TGFB2', 
                 'MMP12',
                 'MMP1']

#The parameters are always overriden by system arguments, otherwise the above params are used.
#The system parameter order is: cell_line (str), filter_crapome (bool), filter_networks_based_on_expression (bool)
if len(sys.argv)>1:
    
    cell_line = sys.argv[1]
    filter_crapome = hf.str_to_bool(sys.argv[2])
    filter_networks_based_on_expression = hf.str_to_bool(sys.argv[3])
    
    
print ('cell_line:',cell_line,
       ' filter_crapome:',filter_crapome, 
       ' filter_networks_based_on_expression:',filter_networks_based_on_expression)
           
#HUBRIS-only (no experimental egdes from our study) ego network of HHIP before filtering
G_ego_HHIP = G_hubris.subgraph(list(G_hubris.neighbors('HHIP'))+['HHIP']).copy()

#filtering based on RNASeq
if filter_networks_based_on_expression:
    
    G_hubris = hf.filter_hubris_based_on_cell_type_specific_gene_list(G_hubris,cell_line,cell_lines,keep_nodes=['HHIP'])  
  
    #taking the LCC
    G_hubris=G_hubris.subgraph(max(nx.connected_components(G_hubris), key=len))
    print('After filtering: N=%d, E=%d'%(G_hubris.number_of_nodes(), G_hubris.number_of_edges())) 

#We load the experimental edges:
df_all_HHIP_links=pd.read_excel('all_significant_links_HHIP.xlsx')
target_proteins={} #the dictionary containing the new edges

#We next determine the new experimental edges based on the input parameters (cell_line, crapome, expression)
#NOTE: there are two 16HBE experiments (06 and 11). We use an OR rule between the links found to be singnificant in either
#, i.e. if a link shows up in either of the two experiments we include it in our network. 
if filter_crapome:
    if cell_line=='IMR90':
        experiment='IMR90_06'
        target_proteins[experiment]=list(df_all_HHIP_links[df_all_HHIP_links['crapome_IMR90_06']==1]['bait_gene_symbol'])
    elif cell_line=='16HBE':
        experiment='16HBE_06_11'
        target_proteins[experiment]=list(df_all_HHIP_links[(df_all_HHIP_links['crapome_16HBE_06']==1) | (df_all_HHIP_links['crapome_16HBE_11']==1)]['bait_gene_symbol'])
    elif cell_line=='union':  
        experiment='union'
        target_proteins[experiment]=list(df_all_HHIP_links[(df_all_HHIP_links['crapome_16HBE_06']==1) | (df_all_HHIP_links['crapome_16HBE_11']==1) | (df_all_HHIP_links['crapome_IMR90_06']==1)]['bait_gene_symbol'])
    elif cell_line=='intersection':
        experiment='intersection'
        target_proteins[experiment]=list(df_all_HHIP_links[((df_all_HHIP_links['crapome_16HBE_06']==1) | (df_all_HHIP_links['crapome_16HBE_11']==1)) & (df_all_HHIP_links['crapome_IMR90_06']==1)]['bait_gene_symbol'])
else:
    if cell_line=='IMR90':
        experiment='IMR90_06'
        target_proteins[experiment]=list(df_all_HHIP_links[df_all_HHIP_links['sainte_IMR90_06']==1]['bait_gene_symbol'])
    elif cell_line=='16HBE':
        experiment='16HBE_06_11'
        target_proteins[experiment]=list(df_all_HHIP_links[(df_all_HHIP_links['sainte_16HBE_06']==1) | (df_all_HHIP_links['sainte_16HBE_11']==1)]['bait_gene_symbol'])
    elif cell_line=='union':  
        experiment='union'
        target_proteins[experiment]=list(df_all_HHIP_links[(df_all_HHIP_links['sainte_16HBE_06']==1) | (df_all_HHIP_links['sainte_16HBE_11']==1) | (df_all_HHIP_links['sainte_IMR90_06']==1)]['bait_gene_symbol'])
    elif cell_line=='intersection':
        experiment='intersection'
        target_proteins[experiment]=list(df_all_HHIP_links[((df_all_HHIP_links['sainte_16HBE_06']==1) | (df_all_HHIP_links['sainte_16HBE_11']==1)) & (df_all_HHIP_links['sainte_IMR90_06']==1)]['bait_gene_symbol'])
                 
if filter_networks_based_on_expression:
    copd_gwas_genes=sorted(list(set(copd_gwas_genes).intersection(set(G_hubris.nodes()))))

#we create a copy of hubris and add the new edges
new_edges=[]
new_edges+=list(zip(['HHIP' for i in range(len(target_proteins[experiment]))], target_proteins[experiment]))
G_hubris_extended=G_hubris.copy()
G_hubris_extended.add_edges_from(new_edges)

#Next we generate the induced graph in each network between HHIP and the other COPD GWAS genes. 
G_induced_hubris,hubris_sp_count_and_len=hf.induced_graph(G_hubris,'HHIP', copd_gwas_genes)
G_induced_hubris_extended,hubris_extended_sp_count_and_len=hf.induced_graph(G_hubris_extended,'HHIP', copd_gwas_genes)

## Plotting the shortest path differences between HUBRIS and HUBRIS extended
plt.rcParams.update({
    "font.family": "Arial",
})

fig, axs=plt.subplots(2,1, figsize=(18,8))
fig.tight_layout()
font = {'family' : 'normal',
        'weight' : 'normal',
        'size'   : 18}
plt.rc('font', **font)
x1=np.arange(0,len(copd_gwas_genes))*3
x2=np.arange(0,len(copd_gwas_genes))*3+1

axs[0].bar(x1,hubris_sp_count_and_len[1,:],label='default')
axs[0].bar(x2,hubris_extended_sp_count_and_len[1,:],label='extended', color='purple')
axs[0].set_title('HUBRIS shortest path lengths in the default vs. extended networks in %s'%cell_line)
axs[0].set_xticks((x1+x2)/2,copd_gwas_genes,fontsize=16)
axs[0].set_ylabel('Shortest path length',fontsize=16)
axs[0].grid()
axs[0].legend()

axs[1].bar(x1,hubris_sp_count_and_len[0,:],label='default',log=True, color='orange')
axs[1].bar(x2,hubris_extended_sp_count_and_len[0,:],label='extended', color='red')
axs[1].set_title('HUBRIS number of shortest paths in the default vs. extended networks in %s'%cell_line)
axs[1].set_xticks((x1+x2)/2,copd_gwas_genes,fontsize=16)
axs[1].set_ylabel('Number of shorthest paths',fontsize=16)
axs[1].grid()
axs[1].legend()
plt.tight_layout()
#axs[0].sharey(axs[1])


plt.savefig('CR_outputs/hubris_sps_cell_type_%s_crapome_%s_expression_%s.png'%(cell_line,str(int(filter_crapome)),str(int(filter_networks_based_on_expression))),dpi=300)
plt.savefig('CR_outputs/hubris_sps_cell_type_%s_crapome_%s_expression_%s.pdf'%(cell_line,str(int(filter_crapome)),str(int(filter_networks_based_on_expression))))

#next we generate the graph that shows the highlighted shortest paths between HHIP and the other GWAS genes

selected_path_targets=np.array(copd_gwas_genes)[(hubris_extended_sp_count_and_len[1,:]<hubris_sp_count_and_len[1,:]) | (hubris_extended_sp_count_and_len[1,:]<=2)]

print('Shortened or <=2 step paths in extened HUBRIS to:',selected_path_targets)

sp_list=[]
for i in selected_path_targets:
    sp_list.append(['HHIP',str(i)])
    
cell_line_column_mapping={'IMR90': set(['sainte_IMR90_06','crapome_IMR90_06',]),
                           '16HBE': set(['sainte_16HBE_06','sainte_16HBE_11','crapome_16HBE_06','crapome_16HBE_11'])}

node_to_cell_line_mapping={}

if filter_crapome:
    relevant_columns = ['bait_gene_symbol', 'crapome_IMR90_06', 'crapome_16HBE_06','crapome_16HBE_11']
else: 
    relevant_columns = ['bait_gene_symbol', 'sainte_IMR90_06', 'sainte_16HBE_06', 'sainte_16HBE_11']

for i,row in df_all_HHIP_links[relevant_columns].iterrows():
    
    if row['bait_gene_symbol'] in target_proteins[experiment]:
        
        columns_where_the_row_is_1=set(row.keys()[(row==1)])
        node_cell_lines=[]
        
        for cl in cell_lines:
            if cell_line_column_mapping[cl].intersection(columns_where_the_row_is_1):
                node_cell_lines.append(cl)
        
        node_label = ' and '.join(node_cell_lines)
        
        node_to_cell_line_mapping[row['bait_gene_symbol']] = node_label

#we decorate the graph for plotting 
G_hubris_seleted_sp=nx.Graph()
for e in sp_list:
    print(e[1])
    print('hubris',list(nx.shortest_paths.all_shortest_paths(G_hubris_extended, e[0],e[1])))

    for sp in list(nx.shortest_paths.all_shortest_paths(G_hubris_extended, e[0],e[1])):
        G_hubris_seleted_sp.add_edges_from(list(zip(sp[:-1],sp[1:])))
        
        
color_scheme={'16HBE':('#1a53ff','#ffffff'), #(node color, font color)
              'IMR90':('#ffcc00','#000000'),
              'IMR90 and 16HBE':('#5ad45a','#000000'),
              'GWAS':('#b30000','#ffffff'),
              'other':('#ffffff','#000000')}

for n in G_hubris_seleted_sp.nodes():
    
    if n in target_proteins[experiment]:
        node_type = node_to_cell_line_mapping[n]
    elif n in copd_gwas_genes:
        node_type='GWAS'
    else:
        node_type='other'
    if n=='HHIP':
        node_type='GWAS'

    G_hubris_seleted_sp.nodes[n]['name']=n
    G_hubris_seleted_sp.nodes[n]['color']=color_scheme[node_type][0]
    G_hubris_seleted_sp.nodes[n]['font_color']=color_scheme[node_type][1]
    G_hubris_seleted_sp.nodes[n]['width']=60
    G_hubris_seleted_sp.nodes[n]['height']=30
    G_hubris_seleted_sp.nodes[n]['font_size']=12    


nx.write_graphml(G_hubris_seleted_sp, 'CR_outputs/G_hubris_seleted_sp_cell_type_%s_crapome_%s_expression_%s.graphml'%(cell_line,str(int(filter_crapome)),str(int(filter_networks_based_on_expression))))


